### ISSUE TITLE
Change the generated code to export a function, not a static machine definition

### ISSUE THREAD

**Issue by mattpocock** [lucydsl/liblucy#34]
https://www.loom.com/share/a1143b7573914ca288765a8c966f70b1

**Comment by matthewp** [comment]
Thanks for raising this issue, I had thought about some sort of function export initially, I was thinking about it in terms of unimplemented actions/guards like you have mentioned. So I was concerned that there might need to be some type of export syntax and a way to define arguments. Now I see that we probably don't need that. Any unimplemented guard/action becomes part of the arguments.

There's probably a lot to discuss here, so. I want to think/talk about this more so we can break down the issue. I would say that the ultimate goal here is to allow completely implementation-free Lucy machine definitions.

Possibly even deprecating/removing implementation touch points like `use` and inline actions/guards. Ergonomics is key here, as inline actions/guards were pretty ergonomic and we don't want to lose that by getting rid of them.

**Comment by matthewp** [comment]
Side-note, this would also fix the fact that you currently cannot define the initial context in Lucy (except clumsily through `@entry` definitions in your initial state).

**Comment by mattpocock** [comment]
> Side-note, this would also fix the fact that you currently cannot define the initial context in Lucy (except clumsily through @entry definitions in your initial state).

Currently, you could do it via `machine.withContext()`. But point taken, a function would be much more ergonomic.

`use` is a really interesting API, because it allows for arbitrary code execution while still allowing the lucy schema to be statically extractable. But typing it might be more trouble than it's worth? Torn.

**Comment by matthewp** [comment]
Some things we need to figure out:

* What named invokes look like.
* The naming convention for exported machine functions (ex. does `ToggleMachine` become `createToggleMachine`? Is that intuitive for the user?)
  * I've also wanted it to be possible to have machine definitions that are *not* exported. How does that work?
* What the arguments should be.
* If there's a more ergonomic way to define named actions/guards than the current form.

**Comment by mattpocock** [comment]
https://www.loom.com/share/8f652e019a9b4eada5ca9662b13cec8e

**Comment by matthewp** [comment]
Nice! Yeah we might be able to just change all of these to function-call style, where the argument is actually the name. Something to think more about, want to look at other DSLs for inspiration here as well.

By non-exported machine definitions I mean something like this:

```lucy
machine uploadMachine {
  // ...
}

machine MainMachine {
  state startup {
    invoke uploadMachine {
      done => idle
    }
  }

  state idle {}
}
```

`uploadMachine` is used internally only. Currently it will be exported though. Ideally we would only export the machines intended to be interpreted. In [Go](https://tour.golang.org/basics/3) the first letter determines if it is exported; if it starts with a capital letter it is, if it starts with a lowercase it is not. Possible we could use that convention too. I'd prefer avoiding `export` keyword if possible (don't think DSLs typically have such a concept).

**Comment by matthewp** [comment]
Types of references in Lucy:

 * Identifiers - References to variables inside file
 * Data - Context properties (idea -> @todo)
 * Implementation names - Named actions/guards/delays, etc. (idea -> :canProceed)
 * Event names - Names of events 

https://www.loom.com/share/2d4297e26e014f22ac218a8faa0a4b23

**Comment by mattpocock** [comment]
@matthewp Nice!

Agree with everything in here. One thing to note is a fifth kind of 'thing' in Lucy - spawned actors in context. In `send(firstArg, secondArg)`, the firstArg is presumably the actor you're targeting? I.e a reference to that actor. So you have references to static definitions, and also references to items in context.

An interesting idea is how much specific actor co-ordination/spawning we want to do within Lucy itself - technically actor spawning is all implementation detail! Since it all lives within context/actions etc.

Also 100% agree that Lucy should be independent from XState. XState is a compile target, and Lucy may end up wanting to support different compile targets. That means that we should really be supporting a subset of what is possible in XState, not looking for feature-completeness.

**Comment by matthewp** [comment]
Spawned actors are saved as properties on the context (ie, data), so they are data like anything else that gets assigned. When you send a message to an actor you are sending it to that context property, so `send(@actorName, eventName)` is correct I think. By allowing Lucy to know about the data in a machine we can statically verify that `actorName` is indeed an *actor* and not some other type of data.

**Comment by mattpocock** [comment]
One thing that faintly worries me is that if you allow 2+ machines inside a file, eventually folks will ask you for modularity, and import/export syntax. Especially when you could conceivably build very complex actor systems out of this. Not necessarily a danger, but something to bear in mind.

**Comment by matthewp** [comment]
I think where I'm landing on this issue is:

1. Change exported machine defintions to be a function that creates the machine. This will make TypeScript easier (future issue).
1. Use a convention of `create[MachineName]` as the exported function name. Single machine Lucy files will continue to export default.
1. Add new syntax: `:name` to signal an external value provided by initialization.

`use` and referencing imported JS will remain in, for now. We can omit showing that in the docs, and assuming all goes well we can deprecate and remove in the future.

Other things discussed in this issue (such as special syntax for data properties) will be done in a future issue.

**Comment by matthewp** [comment]
That works for me.

Using this issue as a cross-repo tracker:

* [ ] - Add symbols to lucy-prism
    * PR: https://github.com/matthewp/lucy-prism/pull/1
* [ ] - Add symbols to lucy-vscode
* [ ] - Update docs to reflect preference for symbol usage and update compilation examples.
* [ ] - Add symbols in liblucy and make sure they work in:
    * [ ] - assign
    * [ ] - action
    * [ ] - guard
    * [ ] - invoke
* [.] - Make invoke callable in liblucy

**Comment by matthewp** [comment]
@mattpocock You mentioned passing `context` through as part of the options. I think that will mean the signature will look something like this:

```js
export default function({ actions, assigns, context, delays, guards, services }) {
  return createMachine({
    context,
    // ...
  });
}
```

My TS skills are a little weak, will we be able to type `context` here using generics?

```typescript
function createToggleMachine<T>({ actions, context }: { actions: SomeType, context: T }) {
```

Is that valid? Would it be valid if the options type here was moved up into an interface as well?

**Comment by mattpocock** [comment]
@matthewp Yes, exactly! Here is the exact type signature you need:

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwKIDcCmA7GB5ARgKzQGMYAaOAWQEMiALYDNbMGYCDAZwF8AzKCEHADkADw4wqMNEIBQMngFcMJNhjhEoaSWgAqEAOb6ANmmp0GaADw6AwuykiycHakzw0jzABMOcV1jxCEgA+AAoIFlUOAC44AAUqWGAqI0szekZmVnYOazssDycXdCxguAAyRHV7QtjbGsc4LmCASkQuIA

**Comment by matthewp** [comment]
This is close to being ready implementation wise. Biggest issue at the moment is what to do when invoking/spawning submachines. How are the child machines going to get their options dependencies? Currently I think we are going to have to say that child machines can't have options deps and we'll just call the `createChildMachine()` function to get the StateMachine instance. Down the line we'll need a solution for this though.

**Comment by matthewp** [comment]
Code is implemented. All is left is updating docs. PR is here: https://github.com/lucydsl/liblucy/pull/41

If anyone wants to review, this is the test to look at:
* Input: https://github.com/lucydsl/liblucy/blob/0de0bf894eccda870761380e2e8a06d0471deea6/test/snapshots/symbols/input.lucy
* Output: https://github.com/lucydsl/liblucy/blob/0de0bf894eccda870761380e2e8a06d0471deea6/test/snapshots/symbols/expected.js

Does this match everyone's expectations? What I like about this is that you can use symbols *anywhere*, in spawn(), in invoke(), not just action/guard.

**Comment by matthewp** [comment]
Closed by #41
